File: tier2_intermediate/bit_manipulation/problems/p0191_number_of_1_bits.cpp
/*
 * P191: Number of 1 Bits (Easy)
 * https://leetcode.com/problems/number-of-1-bits/
 * Topics: Divide and Conquer, Bit Manipulation
 * 
 * Given a positive integer n, write a function that returns the number of set bits in its binary representation (also known as the Hamming weight).
 * 
 * Example 1:
 *     Input: n = 11
 *     Output: 3
 *     Explanation:
 *     The input binary string 1011 has a total of three set bits.
 * 
 * Example 2:
 *     Input: n = 128
 *     Output: 1
 *     Explanation:
 *     The input binary string 10000000 has a total of one set bit.
 * 
 * Example 3:
 *     Input: n = 2147483645
 *     Output: 30
 *     Explanation:
 *     The input binary string 1111111111111111111111111111101 has a total of thirty set bits.
 * 
 * Constraints:
 *     - 1 <= n <= 231 - 1
 * 
 * Template (python3):
 *     class Solution:
 *         def hammingWeight(self, n: int) -> int:
 * 
 * Hint: Use Brian Kernighan's algorithm: n &= (n - 1) clears the lowest set bit.
 */
#include "cpptest.h"
#include <cstdint>

static int hammingWeight(uint32_t n) {
    abort();
}

int main() {
    (void)print_arr;

    struct { const char *label; uint32_t input; int expected; } tests[] = {
        {"example 1", 0b00000000000000000000000000001011u, 3},
        {"power of two", 0b00000000000000000000000010000000u, 1},
        {"leetcode example 3", 2147483645u, 30},
        {"zero", 0u, 0},
        {"single bit", 1u, 1},
        {"alternating bits", 0x55555555u, 16},
        {"max value all bits set", 0x7FFFFFFFu, 31},
    };

    int n = (int)(sizeof(tests) / sizeof(tests[0]));
    int passed = 0;

    printf("\n============================================================\n");
    printf("  191. Number of 1 Bits\n");
    printf("============================================================\n");

    for (int i = 0; i < n; i++) {
        int got = hammingWeight(tests[i].input);
        if (got == tests[i].expected) {
            passed++;
            printf("  Test %d (%s): PASS\n", i + 1, tests[i].label);
        } else {
            printf("  Test %d (%s): FAIL\n", i + 1, tests[i].label);
            printf("    Expected: %d\n    Got:      %d\n", tests[i].expected, got);
        }
    }

    printf("\n  %d/%d passed\n", passed, n);
    printf("============================================================\n\n");

    return passed == n ? 0 : 1;
}

File: tier2_intermediate/bit_manipulation/test_runners/p0191_number_of_1_bits.py
from src.utils.judge_base import JudgeBase, TestCase, register


class P0191NumberOf1Bits(JudgeBase):
    name = "191. Number of 1 Bits"
    test_cases = [
        TestCase(
            input=0b00000000000000000000000000001011,
            expected=3,
            label="example 1",
        ),
        TestCase(
            input=0b00000000000000000000000010000000,
            expected=1,
            label="power of two",
        ),
        TestCase(input=2147483645, expected=30, label="leetcode example 3"),
        TestCase(input=0, expected=0, label="zero"),
        TestCase(input=1, expected=1, label="single bit"),
        TestCase(input=0x55555555, expected=16, label="alternating bits"),
        TestCase(input=(1 << 31) - 1, expected=31, label="max value all bits set"),
    ]

    def to_stdin(self, tc_input) -> str:
        return self.serialize_int(tc_input)

    def check_stdout(self, stdout: str, tc_expected) -> bool:
        return self.parse_int(stdout) == tc_expected


register(P0191NumberOf1Bits)
